### ISSUE TITLE
visualization code

### ISSUE THREAD

**Issue by Encoding101** [xiaomoguhz/OV-DQUO#3]
Hi   @xiaomoguhz , congratulations you work was accpeted AAAI2025. Could provide your image detect visualization code? Thank you！

**Comment by JinJin19991202** [comment]
import torch
import cv2
import numpy as np
import os
import argparse
import warnings
import datetime
import matplotlib.pyplot as plt
from torchvision.ops import nms
from main import build_model_main, get_coco_api_from_dataset, get_args_parser
from datasets import build_dataset
from util.slconfig import DictAction, SLConfig
from models.ov_dquo.ov_postprocess import OVPostProcess
from util.visualizer import COCOVisualizer, renorm
from engine import evaluate  

def parse_args():
    parser = argparse.ArgumentParser(description='Object Detection Visualization', parents=[get_args_parser()])
    parser.add_argument('--image_path', default='data/Images/.jpg',
                        help='Path to the image for visualization')
    parser.add_argument('--conf_threshold', type=float, default=0.5, help='Confidence threshold for filtering boxes')
    parser.add_argument('--nms_threshold', type=float, default=0.5, help='NMS threshold for filtering boxes')
    args = parser.parse_args()
    return args

def get_categories(args):
    dataset_val = build_dataset(image_set="val", args=args)
    categories = dataset_val.category_list
    return categories

def load_model_and_config(args):
    cfg = SLConfig.fromfile(args.config_file)
    if args.options is not None:
        cfg.merge_from_dict(args.options)
    cfg_dict = cfg._cfg_dict.to_dict()
    args_vars = vars(args)
    for k, v in cfg_dict.items():
        if k not in args_vars:
            setattr(args, k, v)
        else:
            raise ValueError(f"Key {k} can be used by args only")

    model, criterion, postprocessors = build_model_main(args)
    device = torch.device(args.device)
    model.to(device)

    checkpoint = torch.load(args.resume, map_location="cpu")
    model.load_state_dict(checkpoint["model"])
    model.eval()

    return model, criterion, postprocessors, device

def prepare_input(image_path, device):
    image = cv2.imread(image_path)  
    if image is None:
        raise FileNotFoundError(f"Image not found at path: {image_path}")

    print(f"Original image shape: {image.shape}")

    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)  
    image_tensor = torch.from_numpy(image_rgb).permute(2, 0, 1).float().unsqueeze(0).to(device) / 255.0

    print(f"Preprocessed image tensor shape: {image_tensor.shape}")

    return image_tensor, torch.tensor([image_rgb.shape[:2]], dtype=torch.float).to(device)

def inference_and_postprocess(model, postprocessors, image, target_sizes, categories, args):
    with torch.no_grad():
        outputs = model(image, categories=categories, targets=None)
    results = postprocessors["bbox"](outputs, target_sizes)

    # Ensure boxes are within the correct range
    boxes = results[0]["boxes"].cpu()
    scores = results[0]["scores"].cpu()
    labels = results[0]["labels"].cpu()
    H, W = target_sizes[0].int().tolist()
    boxes[:, 0::2].clamp_(min=0, max=W)
    boxes[:, 1::2].clamp_(min=0, max=H)

    # Filter boxes by confidence threshold
    keep = scores > args.conf_threshold
    boxes = boxes[keep]
    scores = scores[keep]
    labels = labels[keep]

    # Apply NMS
    keep = nms(boxes, scores, args.nms_threshold)
    boxes = boxes[keep]
    scores = scores[keep]
    labels = labels[keep]

    results[0]["boxes"] = boxes
    results[0]["scores"] = scores
    results[0]["labels"] = labels

    return results

def main():
    args = parse_args()
    model, criterion, postprocessors, device = load_model_and_config(args)
    image, target_sizes = prepare_input(args.image_path, device)
    categories = get_categories(args)
    results = inference_and_postprocess(model, postprocessors, image, target_sizes, categories, args)

    visualizer = COCOVisualizer(categories)
    tgt = {
        "image_id": 1,
        "boxes": results[0]["boxes"].cpu(),
        "size": torch.tensor([image.shape[2], image.shape[3]]),
        "box_label": results[0]["labels"].cpu()
    }

    # Print debug information
    print(f"图像尺寸: {image.shape}")
    print(f"边界框范围: {tgt['boxes'].min()} - {tgt['boxes'].max()}")
    print(f"标签索引: {tgt['box_label']}")
    print(f"标签类别: {[categories[i.item()] for i in tgt['box_label']]}")

    visualizer.visualize(image.squeeze(0), tgt, savedir=args.output_dir)

if __name__ == "__main__":
    warnings.filterwarnings("ignore", category=UserWarning)
    main()

**Comment by Encoding101** [comment]
import torch
import cv2
import numpy as np
import os
import argparse
import warnings
import datetime
import matplotlib.pyplot as plt
from torchvision.ops import nms
from main import build_model_main, get_coco_api_from_dataset, get_args_parser
from datasets import build_dataset
from util.slconfig import DictAction, SLConfig
from models.ov_dquo.ov_postprocess import OVPostProcess
from util.visualizer import COCOVisualizer, renorm
from engine import evaluate  # 从 engine.py 导入评估函数

def parse_args():
    parser = argparse.ArgumentParser(description='Object Detection Visualization', parents=[get_args_parser()])
    parser.add_argument('--image_path', default='data/Images/val2017/000000009400.jpg',
                        help='Path to the image for visualization')
    parser.add_argument('--conf_threshold', type=float, default=0.5, help='Confidence threshold for filtering boxes')
    parser.add_argument('--nms_threshold', type=float, default=0.5, help='NMS threshold for filtering boxes')
    args = parser.parse_args()
    return args


def load_model_and_config(args):
    cfg = SLConfig.fromfile(args.config_file)
    if args.options is not None:
        cfg.merge_from_dict(args.options)
    cfg_dict = cfg._cfg_dict.to_dict()
    args_vars = vars(args)
    for k, v in cfg_dict.items():
        if k not in args_vars:
            setattr(args, k, v)
        else:
            raise ValueError(f"Key {k} can be used by args only")

    model, criterion, postprocessors = build_model_main(args)
    device = torch.device(args.device)
    model.to(device)

    checkpoint = torch.load(args.resume, map_location="cpu")
    model.load_state_dict(checkpoint["model"])
    model.eval()

    return model, criterion, postprocessors, device

def prepare_input(image_path, device):
    image = cv2.imread(image_path)  # BGR顺序
    if image is None:
        raise FileNotFoundError(f"Image not found at path: {image_path}")

    print(f"Original image shape: {image.shape}")

    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)  # 转为RGB
    image_tensor = torch.from_numpy(image_rgb).permute(2, 0, 1).float().unsqueeze(0).to(device) / 255.0

    print(f"Preprocessed image tensor shape: {image_tensor.shape}")

    return image_tensor, torch.tensor([image_rgb.shape[:2]], dtype=torch.float).to(device)

def inference_and_postprocess(model, postprocessors, image, target_sizes, categories, args):
    with torch.no_grad():
        outputs = model(image, categories=categories, targets=None)
    results = postprocessors["bbox"](outputs, target_sizes)

    # Ensure boxes are within the correct range
    boxes = results[0]["boxes"].cpu()
    scores = results[0]["scores"].cpu()
    labels = results[0]["labels"].cpu()
    H, W = target_sizes[0].int().tolist()
    boxes[:, 0::2].clamp_(min=0, max=W)
    boxes[:, 1::2].clamp_(min=0, max=H)

    # Filter boxes by confidence threshold
    keep = scores > args.conf_threshold
    boxes = boxes[keep]
    scores = scores[keep]
    labels = labels[keep]

    # Apply NMS
    keep = nms(boxes, scores, args.nms_threshold)
    boxes = boxes[keep]
    scores = scores[keep]
    labels = labels[keep]

    results[0]["boxes"] = boxes
    results[0]["scores"] = scores
    results[0]["labels"] = labels

    return results


# classes
CLASSES = [
    'N/A', 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus',
    'train', 'truck', 'boat', 'traffic light', 'fire hydrant', 'N/A',
    'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse',
    'sheep', 'cow', 'elephant', 'bear', 'zebra', 'giraffe', 'N/A', 'backpack',
    'umbrella', 'N/A', 'N/A', 'handbag', 'tie', 'suitcase', 'frisbee', 'skis',
    'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove',
    'skateboard', 'surfboard', 'tennis racket', 'bottle', 'N/A', 'wine glass',
    'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple', 'sandwich',
    'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake',
    'chair', 'couch', 'potted plant', 'bed', 'N/A', 'dining table', 'N/A',
    'N/A', 'toilet', 'N/A', 'tv', 'laptop', 'mouse', 'remote', 'keyboard',
    'cell phone', 'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'N/A',
    'book', 'clock', 'vase', 'scissors', 'teddy bear', 'hair drier',
    'toothbrush'
]
print('有对象+无对象', len(CLASSES))


def main():
    args = parse_args()
    model, criterion, postprocessors, device = load_model_and_config(args)
    image, target_sizes = prepare_input(args.image_path, device)
    categories = get_categories(args)
    results = inference_and_postprocess(model, postprocessors, image, target_sizes, categories, args)

  
    visualizer = COCOVisualizer(categories)
    tgt = {
        "image_id": 1,
        "boxes": results[0]["boxes"].cpu(),
        "size": torch.tensor([image.shape[2], image.shape[3]]),
        "box_label": results[0]["labels"].cpu()
    }

    # Print debug information
    print(f"图像尺寸: {image.shape}")
    print(f"边界框范围: {tgt['boxes'].min()} - {tgt['boxes'].max()}")
    print(f"标签索引: {tgt['box_label']}")
    print(f"标签类别: {[categories[i.item()] for i in tgt['box_label']]}")

    visualizer.visualize(image.squeeze(0), tgt, savedir=args.output_dir)

if __name__ == "__main__":
    warnings.filterwarnings("ignore", category=UserWarning)
    main()
